### PR TITLE
Scribe (Android stylus handwriting text input)

### PR DESCRIPTION
**Pull request by justinmc**
Enables the Scribe feature, or Android stylus handwriting text input. This is the bare-minimum feature set, while other features have been scoped out and have POCs (see https://github.com/flutter/flutter/issues/155948, https://github.com/flutter/flutter/issues/156018).

Depends on https://github.com/flutter/flutter/pull/148784

Part of https://github.com/flutter/flutter/issues/115607

scribe

### CODE REVIEW

**Comment by goderbauer** [comment]
(triage): I spoke to @justinmc last week and he says this is still on his radar.

**Comment by justinmc** [comment]
I have a few P1s keeping me from getting back to this, but it's still a priority :+1:

File: shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java
Comment (justinmc) [inline shell/platform/android/io/flutter/plugin/editing/TextInputPlugin.java:360]: Instead of adding this parameter, I could make Scribe's methods part of the TextInputChannel. Then I also wouldn't need to pass ScribeChannel into TextInputPlugin in FlutterView.

However, TextInputChannel is pretty bloated, and it's pretty nice to encapsulate all of the Scribe stuff separately, hence why I did  this.

File: shell/platform/android/test/io/flutter/plugin/editing/ScribePluginTest.java
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/plugin/editing/ScribePluginTest.java:28]: For all of these configs can you set minSdk not sdk. It will run on api 34 and 35 and beyond as we update robolectric.

File: shell/platform/android/test/io/flutter/plugin/editing/ScribePluginTest.java
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/plugin/editing/ScribePluginTest.java:36]: Can you pull the shared logic here out into a setup method?

File: shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java:1396]: same minSdk comment

File: shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java
Comment (justinmc) [reply shell/platform/android/io/flutter/plugin/editing/InputConnectionAdaptor.java:288]: My intention was to indicate that no HandwritingGesture is currently supported. Later I would add support in this method (see https://github.com/flutter/engine/pull/55568).

However I think this doesn't matter because I already didn't register support when I enable stylus handwriting. I'll just remove this method for now until that follow-up PR.

File: shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java
Comment (justinmc) [reply shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java:60]: :+1: I'll add the annotations to ScribeMethodHandler's methods.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java
Comment (justinmc) [reply shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java:33]: If I don't include that I get this warning:

 > warning: [deprecation] setMessageHandler(String,BinaryMessageHandler) in DartExecutor has been deprecated

I got this suppresswarnings line from elsewhere in the codebase like this:

https://github.com/flutter/engine/blob/3388ef926e70c70b44408c165fdfcdeba2164cdc/shell/platform/android/test/io/flutter/plugin/editing/TextInputPluginTest.java#L2007-L2008

File: shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java
Comment (justinmc) [reply shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java:26]: Changed to mInputMethodManager.

File: shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java
Comment (justinmc) [reply shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java:31]: That's the FlutterView from when ScribePlugin is created in FlutterView.java:

```java
      scribePlugin =
          new ScribePlugin(
              this, textInputPlugin.getInputMethodManager(), this.flutterEngine.getScribeChannel());
```

I need it in order to call [InputMethodManager.startStylusHandwriting](https://developer.android.com/reference/android/view/inputmethod/InputMethodManager#startStylusHandwriting(android.view.View)). Is there a better way to get ahold of the View at that time that's called maybe?

Comment (reidbaker) [reply shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java:31]: I am not so confident as to block the pr but I suspect that you can't make that variable final. That there is not a 1 to 1 relationship between flutter view and plugin instantiation. 

I am specifically thinking of add to app and cached engine scenarios. 
https://docs.flutter.dev/add-to-app/android/add-flutter-screen#step-3-optional-use-a-cached-flutterengine

None of our existing tests would catch this because they would not execute this code path.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java
Comment (reidbaker) [reply shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java:37]: Ok I found the java class. Why not use getBinaryMessenger instead? 

https://github.com/flutter/engine/blob/205484009711c2f0d85a4c45666d187184493b41/shell/platform/android/io/flutter/embedding/engine/dart/DartExecutor.java#L224

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/android/FlutterView.java:1143]: I am trying to work out the safety of annotating the plugin with requires 34 and using this code to protect against bad calls. My gut says the plugin should always be created and it should handle the default behavior when running on something lower than 33 but I cant find that in any of our documentation or style guide. 

Additionally the channel is being created in a bunch of code and used in a bunch of code without the api protections. 

Are you confidant that the api methods cannot be called through the channel if we never create the plugin?

File: shell/platform/android/test/io/flutter/plugin/editing/ScribePluginTest.java
Comment (reidbaker) [inline shell/platform/android/test/io/flutter/plugin/editing/ScribePluginTest.java:27]: I missed this on the first pass but please add a test that runs on one of the apis 21-32 and verifies the behavior you expect when running on an android version too old to have scribe. I think with the current implementation it would assert that the ScribePlugin is null.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java
Comment (justinmc) [reply shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java:37]: Is there anywhere in the codebase that's using this that I can use as an example? I'm struggling to get it to work.

I saw something kind of similar here: https://github.com/flutter/engine/blob/d38f5e560a98f491fb1b3e24e19cc78d2b110b99/shell/platform/android/test/io/flutter/embedding/engine/dart/DartExecutorTest.java#L55

And then trying to use getBinaryMessenger like this, but the call never reaches ScribeChannel:

```java
    ArgumentCaptor<BinaryMessenger.BinaryMessageHandler> binaryMessageHandlerCaptor =
        ArgumentCaptor.forClass(BinaryMessenger.BinaryMessageHandler.class);
    FlutterJNI fakeFlutterJni = mock(FlutterJNI.class);
    mockDartExecutor = new DartExecutor(fakeFlutterJni, mock(AssetManager.class));
    //mockDartExecutor = mock(DartExecutor.class);
    assertNotNull(mockDartExecutor.getBinaryMessenger());
    mockHandler = mock(ScribeChannel.ScribeMethodHandler.class);
    ScribeChannel scribeChannel = new ScribeChannel(mockDartExecutor);

    scribeChannel.setScribeMethodHandler(mockHandler);

    MethodCall methodCall = new MethodCall(ScribeChannel.METHOD_START_STYLUS_HANDWRITING, null);
    ByteBuffer encodedMethodCall = StandardMethodCodec.INSTANCE.encodeMethodCall(methodCall);
    mockDartExecutor.getBinaryMessenger().send(ScribeChannel.METHOD_START_STYLUS_HANDWRITING, encodedMethodCall);

    verify(mockHandler).startStylusHandwriting();
```

Comment (reidbaker) [reply shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java:37]: I dont think you want to make a new binary messenger but you want to get one. There is a getBinaryMessenger on the flutterview class. 

For examples maybe this is helpful. https://github.com/flutter/engine/blob/9fa363bdde2085893f5802890ef4ce67753dfd77/shell/platform/android/io/flutter/embedding/android/KeyboardManager.java#L122

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (reidbaker) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: For simplicity we can say that flutter supports the feature on api 34 and above.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java
Comment (justinmc) [reply shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java]: After an unsuccessful deep dive into getBinaryMessenger I was able to work around this just by casting the DartExecutor to BinaryMessenger before mocking setMessageHandler. MethodChannel uses BinaryMessenger.setMessageHandler ([here](https://github.com/flutter/engine/blob/ee926283f8a177d3f4dc994b6e20d24bd94b4b17/shell/platform/android/io/flutter/plugin/common/MethodChannel.java#L143-L144)), which is what I want the mock to catch, and it's not deprecated.

```diff
-    verify(mockBinaryMessenger, times(1))
+    verify((BinaryMessenger) mockBinaryMessenger, times(1))
        .setMessageHandler(any(String.class), binaryMessageHandlerCaptor.capture());
```

File: shell/platform/android/io/flutter/embedding/android/FlutterView.java
Comment (justinmc) [reply shell/platform/android/io/flutter/embedding/android/FlutterView.java]: I've removed this `if` so that ScribePlugin is still instantiated on older SDK versions, but the methods themselves are locked down to the correct API level. I also added tests for unsupported API levels as @reidbaker mentioned in another comment.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java
Comment (justinmc) [inline shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java:89]: I would have liked to test the exact argument that `reply` was called with here, but I was struggling to do anything useful with the ByteBuffer. I was trying `StandardMethodCodec.INSTANCE.decodeEnvelope` because it looks like that's what it's encoded with (e.g. [here](https://github.com/flutter/engine/blob/fd6082ab669356ae1525adc809cc0d0c78109b63/shell/platform/android/io/flutter/plugin/common/MethodChannel.java#L277)), but I always got an underflow error.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/ScribeChannel.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/embedding/engine/systemchannels/ScribeChannel.java:67]: Does this mean that the dart side of the plugin needs to check if it is running on android and check the api level?

File: shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java
Comment (reidbaker) [inline shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java:30]: I think you want this to be private or at least public and @visibleForTesting not public. 
Then add a setter for the view. My gut is that over time if the view changes then you want to be informed and be able to setup new callbacks or other events.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java
Comment (reidbaker) [reply shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java:89]: Instead of trying to reverse engineer the reply bytebuffer what if you used a spy on the result callback and verified that either callback.error or result.error was called. 

I do not love that the error condition verifies the same thing as the success condition but at least we verify we do not call the methods we are trying to protect.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/ScribeChannel.java
Comment (justinmc) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/ScribeChannel.java:67]: I guess in practice this probably means that you want to call this wrapped in a `try/catch` in Dart. My thought process is that this method channel method is just a proxy for InputMethodManager.isStylusHandwritingAvailable. If I'm not even able to call that, I should error, rather than succeed with `false`, which the app developer might interpret to mean that InputMethodManager.isStylusHandwritingAvailable returned false.

If that sounds reasonable then I'll at least make sure this is clearly documented in the framework.

File: shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java
Comment (justinmc) [reply shell/platform/android/test/io/flutter/embedding/engine/systemchannels/ScribeChannelTest.java]: I changed this to JSONMethodCodec which I believe is what we're supposed to be using for these kinds of method channels now, and incidentally I was able to decode and test the value of `result` then.

File: shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java
Comment (justinmc) [reply shell/platform/android/io/flutter/plugin/editing/ScribePlugin.java:32]: (From another thread) I've made a view setter so it can be updated if needed.

**Comment by justinmc** [comment]
I think I've come around to Reid's approach. Here's what I've done:

  * Added `isFeatureAvailable`. I left off the "Scribe" because it felt redundant.
  * Made it call over to `isStylusHandwritingAvailable`, so from the Dart side you only need to call one method.

I think it's important to match the underlying Android APIs, which I've done with isStylusHandwritingAvailable and startStylusHandwriting, but I'm not opposed to adding in a convenience method where it makes sense. In this case it makes the Dart side much easier.

This is what the Dart code would have looked like without isFeatureAvailable:

```dart
try {
  if (!(await Scribe.isStylusHandwritingAvailable() ?? false)) {
    // If isStylusHandwritingAvailable returns false then the device's API level
    // supports Scribe, but for some other reason it's not able to accept stylus
    // input right now.
    return;
  }
} on PlatformException catch (exception) {
  if (exception.message == 'Requires API level 34 or higher.') {
    // The device's API level is too low to support Scribe. It's unfortunate
    // that we have to rely on matching this string from the engine.
    return;
  }
  // Any other exception is unexpected and should not be caught here.
  rethrow;
}

// Scribe is supported, so start it.
Scribe.startStylusHandwriting();
```

With isFeatureAvailable it looks like this:

```dart
if (!(await Scribe.isFeatureAvailable() ?? false)) {
  // The device doesn't support stylus input right now, or maybe at all.
  return;
}

// Scribe is supported, so start it.
Scribe.startStylusHandwriting();
```

For users that need to use the Android API directly for more nuance, they can use isStylusHandwritingAvailable.

File: shell/platform/android/io/flutter/embedding/engine/systemchannels/ScribeChannel.java
Comment (justinmc) [reply shell/platform/android/io/flutter/embedding/engine/systemchannels/ScribeChannel.java:81]: Update here: https://github.com/flutter/engine/pull/52943#issuecomment-2423177793

**Review by gmackall — APPROVED** [review]
LGTM, on board with the addition of `isFeatureAvailable` for convenience - it seems like it will make the end use in dart a good bit nicer

File: shell/platform/android/test/io/flutter/plugin/editing/ScribePluginTest.java
Comment (reidbaker) [reply shell/platform/android/test/io/flutter/plugin/editing/ScribePluginTest.java:81]: I think you *should* keep the test it just seems to me that the error would be something like "feature not available" or "api too low" or something. "No such method". That type of error feels like a programing mistake not an error that tells you what is wrong.

**Comment by reidbaker** [comment]
I really like the change to the dart code that comes with isFeatureAvailable. @justinmc you rock!
https://github.com/flutter/engine/pull/52943#issuecomment-2423177793

**Comment by justinmc** [comment]
There was a failure that seems irrelevant and I'm hoping will go away after I push a merge commit.

```
[ +189 ms] org-dartlang-app:///main.dart:1:1: Error: The specified language version is too high. The highest supported language version is 3.6.
[        ] // @dart = 3.7
```

**Pull request merged**
